From 551b4d8c826e9863abb2e9ab595b5bb996dc2ae8 Mon Sep 17 00:00:00 2001
From: Himal Prasad Ghimiray <himal.prasad.ghimiray@intel.com>
Date: Mon, 3 Aug 2026 10:39:57 +0530
Subject: drm/xe/bo: account kernel pinned vram

Track kernel pinned VRAM in the per-region total counter on pin
and unpin.

Reviewed-by: Tejas Upadhyay tejas.upadhyay@intel.com
Signed-off-by: Himal Prasad Ghimiray <himal.prasad.ghimiray@intel.com>
---
 src/drivers/gpu/drm/xe/xe_bo.c | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/drivers/gpu/drm/xe/xe_bo.c b/drivers/gpu/drm/xe/xe_bo.c
index bef139ed4..ec698b93d 100644
--- a/drivers/gpu/drm/xe/xe_bo.c
+++ b/drivers/gpu/drm/xe/xe_bo.c
@@ -3236,7 +3236,15 @@ int xe_bo_pin(struct xe_bo *bo, struct drm_exec *exec)
 		return err;
 
 	if (mem_type_is_vram(place->mem_type) || bo->flags & XE_BO_FLAG_GGTT) {
+#ifdef BPM_DRMM_CGROUP_REGISTER_REGION_NOT_PRESENT
+		struct xe_ttm_vram_mgr *mgr = xe_bo_pin_vram_mgr(bo);
+#endif
+
 		spin_lock(&xe->pinned.lock);
+#ifdef BPM_DRMM_CGROUP_REGISTER_REGION_NOT_PRESENT
+		if (mgr)
+			mgr->pin.total_used += xe_bo_size(bo);
+#endif
 		if (bo->flags & XE_BO_FLAG_PINNED_LATE_RESTORE)
 			list_add_tail(&bo->pinned_link, &xe->pinned.late.kernel_bo_present);
 		else
@@ -3259,6 +3267,7 @@ int xe_bo_pin(struct xe_bo *bo, struct drm_exec *exec)
 
 /**
  * xe_bo_unpin_external - unpin an external BO
+#endif
  * @bo: buffer object to be unpinned
  *
  * Unpin an external (not tied to a VM, can be exported via dma-buf / prime FD)
@@ -3316,7 +3325,14 @@ void xe_bo_unpin(struct xe_bo *bo)
 	xe_assert(xe, xe_bo_is_pinned(bo));
 
 	if (mem_type_is_vram(place->mem_type) || bo->flags & XE_BO_FLAG_GGTT) {
+#ifdef BPM_DRMM_CGROUP_REGISTER_REGION_NOT_PRESENT
+		struct xe_ttm_vram_mgr *mgr = xe_bo_pin_vram_mgr(bo);
+#endif
 		spin_lock(&xe->pinned.lock);
+#ifdef BPM_DRMM_CGROUP_REGISTER_REGION_NOT_PRESENT
+		if (mgr)
+			mgr->pin.total_used -= xe_bo_size(bo);
+#endif
 		xe_assert(xe, !list_empty(&bo->pinned_link));
 		list_del_init(&bo->pinned_link);
 		spin_unlock(&xe->pinned.lock);
-- 
2.43.0

